File: Qt e Progetti/Progetti/2017-2018 Progetto Calcolatrice/Polygonator - Menon e Dalla Via/Java/src/Model/Polygons/Triangle.java
package Model.Polygons;

import Model.Line;
import Model.Point;

import java.util.ArrayList;
import java.util.List;

public class Triangle extends Polygon {
    public Triangle(final List<Point> vertices) throws Exception {
        super(vertices);
    }

    /**
     * Returns whether a list of points can make a triangle or not
     */
    public static boolean canBeCreatedWith(final Polygon polygon) {
        return polygon.getVertices().size() == 3;
    }

    @Override
    public double getApothemConstant() {
        return 0.289;
    }

    @Override
    public List<Line> getDiagonals() throws Exception {
        throw new Exception("A triangle does not have diagonals");
    }

    /**
     * Returns the height of the calculated base of the triangle
     */
    @Override
    public double getHeight() throws Exception {
        if(getVertices().get(0).getY() >= getVertices().get(1).getY()) {
            return Line.getDistance(vertices.get(1), vertices.get(2), vertices.get(0));
        }

        return Line.getDistance(vertices.get(0), vertices.get(1), vertices.get(2));
    }

    /**
     * Returns the area of a triangle
     */
    @Override
    public double getArea() throws Exception {
        return getBase() * getHeight() / 2;
    }

    /**
     * Returns the list of all the medians of a triangle
     */
    public List<Line> getMedians() throws Exception {
        List<Line> medians = new ArrayList<>();

        for (Point vertex : vertices) {
            medians.add(getMedian(vertex));
        }

        return medians;
    }

    /**
     * Returns the median between the given point and the opposite line
     */
    public Line getMedian(final Point p) throws Exception {
        Point a = p.equals(vertices.get(0)) ? vertices.get(1) : vertices.get(0);
        Point b = p.equals(vertices.get(2)) ? vertices.get(1) : vertices.get(2);

        return new Line(Line.getMidPoint(a, b), p);
    }

    /**
     * Returns the list of all the bisectors of the triangle
     */
    public List<Line> getBisectors() throws Exception {
        List<Line> bisectors = new ArrayList<>();

        for (Point vertex : vertices) {
            bisectors.add(getBisector(vertex));
        }

        return bisectors;
    }

    /**
     * Returns the bisector of the given vertex
     */
    public Line getBisector(final Point p) throws Exception {
        Point a = p.equals(vertices.get(0)) ? vertices.get(1) : vertices.get(0);
        Point b = p.equals(vertices.get(2)) ? vertices.get(1) : vertices.get(2);

        double ap = Line.getDistance(a, p);
        double bp = Line.getDistance(b, p);

        double x = (ap * b.getX() + bp * a.getX()) / (ap + bp);
        double y = (ap * b.getY() + bp * a.getY()) / (ap + bp);

        return new Line(new Point(x, y), p);
    }

    @Override
    public String toString() {
        return "Triangle";
    }
}

File: Qt e Progetti/Progetti/2017-2018 Progetto Calcolatrice/Polygonator - Menon e Dalla Via/C++/Model/Polygons/Triangle.cpp
#include "Triangle.h"

Triangle::Triangle(const list<Point> &vertices) : Polygon(vertices) {}

bool Triangle::canBeCreatedWith(const Polygon &polygon) {
    return polygon.getVertices().size() == 3;
}

double Triangle::getApothemConstant() const {
    return 0.289;
}

list<Line> Triangle::getDiagonals() const {
    throw Exception("A triangle does not have diagonals");
}

double Triangle::getHeight() const {
    if( (*getVertices().begin()).getY() >= (*next(next(getVertices().begin()))).getY() ) {
        return Line::getDistance(*next(getVertices().begin()), *next(next(getVertices().begin())), *getVertices().begin());
    }

    return Line::getDistance(*getVertices().begin(), *next(getVertices().begin()), *prev(getVertices().end()));
}

double Triangle::getArea() const {
    return getBase() * getHeight() / 2;
}

list<Line> Triangle::getMedians() const {
    list<Line> medians;

    for (Point vertex : getVertices()) {
        medians.push_back(getMedian(vertex));
    }

    return medians;
}

Line Triangle::getMedian(const Point &p) const {
    Point a = *getVertices().begin() == p ? *next(getVertices().begin()) : *getVertices().begin();
    Point b = *prev(getVertices().end()) == p ? *next(getVertices().begin()) : *prev(getVertices().end());

    return {Line::getMidPoint(a, b), p};
}

list<Line> Triangle::getBisectors() const {
    list<Line> bisectors;

    for (Point vertex : getVertices()) {
        bisectors.push_back(getBisector(vertex));
    }

    return bisectors;
}

Line Triangle::getBisector(const Point &p) const {
    Point a = *getVertices().begin() == p ? *next(getVertices().begin()) : *getVertices().begin();
    Point b = *prev(getVertices().end()) == p ? *next(getVertices().begin()) : *prev(getVertices().end());

    double ap = Line::getDistance(a, p);
    double bp = Line::getDistance(b, p);

    double x = (ap * b.getX() + bp * a.getX()) / (ap + bp);
    double y = (ap * b.getY() + bp * a.getY()) / (ap + bp);

    return {{x, y}, p};
}

string Triangle::getName() const {
    return "Triangle";
}